File: Algorithms/src/2092. Find All People With Secret/Code/solution.py
class Solution:
    def findAllPeople(self, n, meetings, firstPerson):
        meetings.sort(key=lambda x: x[2])

        parent = list(range(n))
        knows = [False] * n
        knows[0] = knows[firstPerson] = True

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            x, y = find(x), find(y)
            if x != y:
                parent[y] = x

        i = 0
        while i < len(meetings):
            time = meetings[i][2]
            people = []

            j = i
            while j < len(meetings) and meetings[j][2] == time:
                union(meetings[j][0], meetings[j][1])
                people.extend([meetings[j][0], meetings[j][1]])
                j += 1

            good = set(find(p) for p in people if knows[p])

            for p in people:
                if find(p) in good:
                    knows[p] = True
                else:
                    parent[p] = p

            i = j

        return [i for i in range(n) if knows[i]]

File: Algorithms/src/2092. Find All People With Secret/Code/solution.go
func findAllPeople(n int, meetings [][]int, firstPerson int) []int {
    sort.Slice(meetings, func(i, j int) bool {
        return meetings[i][2] < meetings[j][2]
    })

    parent := make([]int, n)
    for i := range parent {
        parent[i] = i
    }

    knows := make([]bool, n)
    knows[0], knows[firstPerson] = true, true

    var find func(int) int
    find = func(x int) int {
        if parent[x] != x {
            parent[x] = find(parent[x])
        }
        return parent[x]
    }

    union := func(x, y int) {
        x, y = find(x), find(y)
        if x != y {
            parent[y] = x
        }
    }

    i := 0
    for i < len(meetings) {
        time := meetings[i][2]
        people := []int{}

        j := i
        for j < len(meetings) && meetings[j][2] == time {
            union(meetings[j][0], meetings[j][1])
            people = append(people, meetings[j][0], meetings[j][1])
            j++
        }

        good := map[int]bool{}
        for _, p := range people {
            if knows[p] {
                good[find(p)] = true
            }
        }

        for _, p := range people {
            if good[find(p)] {
                knows[p] = true
            } else {
                parent[p] = p
            }
        }
        i = j
    }

    res := []int{}
    for i := 0; i < n; i++ {
        if knows[i] {
            res = append(res, i)
        }
    }
    return res
}
